src/OrchardCore/Orchard.DeferredTasks/DeferredTaskMiddleware.cs

```diff
@@ -30,16 +30,27 @@ public DeferredTaskMiddleware(RequestDelegate next, IShellHost orchardHost)
             // We only serve the next request if the tenant has been resolved.
             if (shellSettings != null)
             {
-                var shellContext = _orchardHost.GetOrCreateShellContext(shellSettings);
+                var deferredTaskEngine = httpContext.RequestServices.GetService<IDeferredTaskEngine>();
 
-                using (var scope = shellContext.CreateServiceScope())
+                // Create a new scope only if there are pending tasks
+                if (deferredTaskEngine != null && deferredTaskEngine.HasPendingTasks)
                 {
-                    var deferredTaskEngine = scope.ServiceProvider.GetService<IDeferredTaskEngine>();
+                    // Dispose the scoped services for the current request, and create a new one
+                    (httpContext.RequestServices as IDisposable).Dispose();
 
-                    if (deferredTaskEngine != null && deferredTaskEngine.HasPendingTasks)
+                    var shellContext = _orchardHost.GetOrCreateShellContext(shellSettings);
+
+                    if (!shellContext.Released)
```

Comment:
Oh yes, you mean that the newly created shellContext could have been used and released by another request. So, maybe other pending tasks will be executed but on another context.